examples/fft.dx

```diff
@@ -0,0 +1,219 @@
+on the [Fhutark demo](https://github.com/diku-dk/fft/blob/master/lib/github.com/diku-dk/fft/stockham-radix-2.fut#L30).
+For non-power-of-2 sized arrays, it uses
+[Bluestein's Algorithm](https://en.wikipedia.org/wiki/Chirp_Z-transform),
+which calls the power-of-2 FFT as a subroutine.
+
+
+'## General helper functions
+
+def isPowerOf2 (x:Int) : Bool =
+  -- A fast trick based on bitwise AND.
+  -- Note: The bitwise and operator (.&.)
+  -- is only defined for Byte, which is why
+  -- I use %and here.
+  -- A test below verifies that this works on
+  -- more than 8 bit integers.
+  -- TODO: Make (.&.) polymorphic.
+  if x == 0
+    then False
+    else 0 == %and x (x - 1)
+
+def intpow [Mul a] (base:a) (power:Int) : a =
+  -- Todo: Use Knuth's power trick,
+  -- which can give an asymptotic speedup.
+  reduce one (*) (for a:(Fin power). base)
+
+def intlog2 (x:Int) : Int =
+  yieldState (-1) \ansRef.
+    runState 1 \cmpRef.
+      while do
+        if x >= (get cmpRef)
+          then 
+            ansRef := (get ansRef) + 1
+            cmpRef := %shl (get cmpRef) 1
+            True
+          else
+            False
+
+def intpow2 (power:Int) : Int = %shl 1 power
+
+def nextpow2 (x:Int) : Int = case isPowerOf2 x of
+  True -> x
+  False -> intpow2 (1 + intlog2 x)
+
+def reflect (i:n) : n =
+  unsafeFromOrdinal n ((size n) - 1 - ordinal i)
+
+def listToTable ((AsList n xs): List a) : (Fin n)=>a = xs
+
+def odd_sized_palindrome (mid:a) (seq:n=>a) :
+  ({backward:n | mid:Unit | zforward:n}=>a) =  -- Alphabetical order matters here.
+  -- Turns sequence 12345 into 543212345.
+  for i.
+    case i of
+      {|backward=i|} -> seq.(reflect i)
+      {|mid=i|} -> mid
+      {|zforward=i|} -> seq.i
+
+
+'## Inner FFT functions
+
+data FTDirection =
+  ForwardFT
+  InverseFT
+
+def butterfly_ixs (j':halfn) (pow2:Int) : (n & n & n & n) =
+  -- Re-index at a finer frequency.
+  -- halfn must have half the size of n.
+  -- For explanation, see https://en.wikipedia.org/wiki/Butterfly_diagram
+  -- Note: with fancier index sets, this might be replacable by reshapes.
+  j = ordinal j'
+  k = ((idiv j pow2) * pow2 * 2) + mod j pow2
+  left_write_ix  = unsafeFromOrdinal n k
+  right_write_ix = unsafeFromOrdinal n (k + pow2)
+
+  left_read_ix  = unsafeFromOrdinal n j
+  right_read_ix = unsafeFromOrdinal n (j + size halfn)
+  (left_read_ix, right_read_ix, left_write_ix, right_write_ix)
+
+def power_of_2_fft (direction: FTDirection) (x: n=>Complex) : n=>Complex =
+  -- Input size must be a power of 2.
+  -- Can enforce this with tables-as-index-sets like:
+  -- (x: (log2n=>(Fin 2))=>Complex)) once that's supported.
+  dir_const = case direction of
+    ForwardFT -> -pi
+    InverseFT -> pi
+
+  log2n = intlog2 (size n)
+  halfn = idiv (size n) 2
+  
+  ans = yieldState x \xRef.
+    for i:(Fin log2n).
+      ipow2 = intpow 2 (ordinal i)
+      xRef := yieldAccum (AddMonoid Complex) \bufRef.
+        for j:(Fin halfn).  -- Executes in parallel.
```

Comment:
You wish 😆 It reads from a stateful reference, so I doubt we would ever try to parallelize this. Of course we could (and should) optimize it so that reads are ok, but it's not implemented at the moment!

Reply:
Sorry I think it's an effect of me advocating for moving the `copy` part into the loop 😕